File: LeetCode/904_FruitIntoBaskets/Solution.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        unordered_map<int, int> map;
        int start=0;
        int ret = 0;
        int size = fruits.size();
        for(int i=0;i<size;i++) {
            int fruit = fruits[i];
            if(map.find(fruit)==map.end()){
                map.insert({fruit,1});
            }
            else{
                map[fruit]++;
            }
            while(map.size()>2){
                map[fruits[start]]--;
                if(map[fruits[start]]==0) map.erase(fruits[start]);
                start++;
            }
            if(ret<i-start+1){
                ret = i-start+1;
            }
        }
        return ret;
    }
};

File: LeetCode/904_FruitIntoBaskets/Solution.kt
class Solution {
    fun totalFruit(fruits: IntArray): Int {
        var ret=0
        val table=IntArray(fruits.size)
        var left=0
        var basket=0

        for(right in fruits.indices) {
            if(table[fruits[right]]==0)basket++
            table[fruits[right]]++
            while(basket>2){
                table[fruits[left]]--
                if (table[fruits[left]]==0)basket--
                left++
            }
            ret=maxOf(ret,right-left+1)
        }
        return ret
    }
}
